crates/libs/core/src/runtime/stateful_types.rs

```diff
@@ -129,88 +133,134 @@ impl From<FABRIC_REPLICA_STATUS> for ReplicaStatus {
+        for i in info_iter {
+            // 2 vec has the same length, this cannot fail
+            let ex = ex1_iter.next().unwrap();
+            i.Reserved = ex as *const FABRIC_REPLICA_INFORMATION_EX1 as *mut c_void;
+        }
+
+        let config = FABRIC_REPLICA_SET_CONFIGURATION {
+            ReplicaCount: replica_raw_cache.len() as u32,
+            Replicas: replica_raw_cache.as_ptr(),
             WriteQuorum: self.WriteQuorum,
             Reserved: std::ptr::null_mut(),
+        };
+
+        ReplicaSetConfigView {
+            _replica_raw_vec: replica_raw_cache,
+            _replica_ex1_vec: replica_ex1_cache,
+            raw: config,
+            _phantom: PhantomData,
         }
     }
+}
 
-    fn fill_cache_copy(&mut self) {
-        self.replica_raw_cache.clear();
-        for replica in &self.Replicas {
-            self.replica_raw_cache.push(replica.get_raw());
-        }
+// View is not movable because it has raw pointer links in it.
+// And it has the same lifetime as the config.
+pub struct ReplicaSetConfigView<'a> {
+    _replica_raw_vec: Vec<FABRIC_REPLICA_INFORMATION>,
+    _replica_ex1_vec: Vec<FABRIC_REPLICA_INFORMATION_EX1>,
+    raw: FABRIC_REPLICA_SET_CONFIGURATION,
+    _phantom: PhantomData<&'a ReplicaSetConfig>,
+}
+
+impl ReplicaSetConfigView<'_> {
+    // returns the config that can be passed to SF com api.
+    pub fn get_raw(&self) -> &FABRIC_REPLICA_SET_CONFIGURATION {
+        &self.raw
     }
 }
 
-#[derive(Debug)]
+// Safe wrapping for FABRIC_REPLICA_INFORMATION
+#[derive(Debug, PartialEq, Clone)]
 pub struct ReplicaInfo {
     pub Id: i64,
     pub Role: Role,
     pub Status: ReplicaStatus,
     pub ReplicatorAddress: ::windows_core::HSTRING,
     pub CurrentProgress: i64,
     pub CatchUpCapability: i64,
-    //pub Reserved: *mut ::core::ffi::c_void,
+    pub MustCatchUp: bool,
 }
 
 impl From<&FABRIC_REPLICA_INFORMATION> for ReplicaInfo {
     fn from(r: &FABRIC_REPLICA_INFORMATION) -> Self {
+        let ex1 = r.Reserved as *const FABRIC_REPLICA_INFORMATION_EX1;
+        let mut must_catchup = false;
+        if !ex1.is_null() {
+            if let Some(ex1ref) = unsafe { ex1.as_ref() } {
+                must_catchup = ex1ref.MustCatchup.as_bool();
+            }
+        }
         ReplicaInfo {
             Id: r.Id,
             Role: r.Role.into(),
             Status: r.Status.into(),
-            ReplicatorAddress: HSTRING::from_wide(unsafe { r.ReplicatorAddress.as_wide() })
-                .unwrap(),
+            ReplicatorAddress: HSTRINGWrap::from(r.ReplicatorAddress).into(),
             CurrentProgress: r.CurrentProgress,
             CatchUpCapability: r.CatchUpCapability,
+            MustCatchUp: must_catchup,
         }
     }
 }
 
 impl ReplicaInfo {
-    pub fn get_raw(&self) -> FABRIC_REPLICA_INFORMATION {
-        FABRIC_REPLICA_INFORMATION {
+    // The parts has the same lifetime as self.
+    // Caller needs to stitch the parts together.
```

Comment:
nit: What does "stitch the parts together" mean here? From background conversations, FABRIC_REPLICA_INFORMATION::Reserved needs to point at FABRIC_REPLICA_INFORMATION_EX1, but would not be immediately obvious to a fresh reader

Reply:
Added the details.